File: bwu/string/383-ransom-note.py
class Solution(object):
    def canConstruct(self, ransomNote, magazine):
        """
        :type ransomNote: str
        :type magazine: str
        :rtype: bool
        """
        dict = {}
        for c in magazine:
            if c not in dict:
                dict[c] = 1
            else:
                dict[c] += 1
        
        for c in ransomNote:
            if c not in dict or dict[c] == 0:
                return False
            else:
                dict[c] -= 1
                
        return True

File: captainhcg/go/383-ransom-note.go
func canConstruct(ransomNote string, magazine string) bool {
    mmap := makeMap(magazine)
    nmap := makeMap(ransomNote)
    for ch := range nmap{
        cnt, valid := mmap[ch]
        if valid == false{
            return false
        }
        if cnt < nmap[ch]{
            return false
        }
    }
    return true
}

func makeMap(s string) map[rune]int{
    m := make(map[rune]int)
    for _, ch := range s{
        cnt, valid := m[ch]
        if valid{
            m[ch] = cnt + 1
        } else {
            m[ch] = 1
        }
    }
    return m
}
